knowledge/ios/MASVS-CRYPTO/MASTG-KNOW-0070.md

```diff
@@ -2,8 +2,98 @@
+
+```bash
+frida-trace -n 'MASTestApp' -i "*FixedWidthInteger*random*"
+  ...
+  2959 ms  $ss17FixedWidthIntegerPsE6random2inxSNyxG_tFZ()
+  2959 ms     | arc4random_buf(buf=0x16ef965a8, nbytes=0x8)
+```
+
+Therefore, using the Swift standard library's random APIs with the default `SystemRandomNumberGenerator` is generally suitable for cryptographic purposes on Apple platforms because that generator is defined to use a cryptographically secure algorithm backed by the system CSPRNG.
+
+**Note:** The API also offers additional overloads that accept a custom random number generator conforming to the `RandomNumberGenerator` protocol. For example, the previous `UInt8.random(in: 0...255)` is an alias for:
+
+```swift
+var rng = SystemRandomNumberGenerator()
+let b = UInt8.random(in: 0...255, using: &rng)
+```
+
+But developers can implement their own `RandomNumberGenerator`, which may not be secure. Therefore, when using custom generators, ensure they are suitable for cryptographic use cases. See [SE-0202](https://github.com/swiftlang/swift-evolution/blob/main/proposals/0202-random-unification.md) for more details.
+
+## CommonCrypto
+
+The CommonCrypto library provides the `CCRandomGenerateBytes` and `CCRandomCopyBytes` functions for generating cryptographically secure random bytes. While these functions are secure, they are lower-level APIs compared to `SecRandomCopyBytes` and require more careful handling by developers.
+
+## /dev/random
+
+Direct use of `/dev/random` via `open` and `read` is discouraged because it is a low level interface that is easy to misuse from application code. On Apple platforms, `/dev/random` and `/dev/urandom` are backed by the same Fortuna based kernel CSPRNG and behave equivalently, so the usual Linux advice about `/dev/random` blocking when entropy is low does not apply here. For iOS apps, Apple recommends using higher level APIs such as `SecRandomCopyBytes` or the Swift standard library random APIs instead of reading these device files directly.
+
+## arc4random
+
+The `arc4random` family of functions (`arc4random()`, `arc4random_buf()`, `arc4random_uniform()`) is also available on iOS. On modern Apple platforms these functions are backed by the same kernel CSPRNG as `SecRandomCopyBytes`, and are suitable for cryptographic use, but they are legacy C style interfaces and are easier to misuse than the Swift standard library or `SecRandomCopyBytes`. For example:
+
+- Using `arc4random() % n` to generate a bounded value can introduce [modulo bias](https://en.wikipedia.org/wiki/Fisher%E2%80%93Yates_shuffle#Modulo_bias), where some outcomes are slightly more likely than others.
+- `arc4random_uniform(n)` is specifically designed to avoid modulo bias for arbitrary upper bounds, and should be preferred over `arc4random() % n`.
+- `arc4random_buf()` produces cryptographically strong random bytes, but requires manual buffer management and error handling.
```

Comment:
```suggestion
- Using `arc4random() % n` to generate a bounded value can introduce [modulo bias](https://en.wikipedia.org/wiki/Fisher%E2%80%93Yates_shuffle#Modulo_bias), where some outcomes are slightly more likely than others.
- `arc4random_buf()` produces cryptographically strong random bytes into a caller provided buffer, and does not itself suffer from modulo bias. Any bias only appears if you convert those bytes to a bounded range incorrectly, for example by doing `% n` on derived integers.
- `arc4random_uniform(n)` is specifically designed to avoid modulo bias for arbitrary upper bounds, returning a uniformly distributed integer in the range `[0, n)`, and should be preferred over `arc4random() % n`.
```